### PR TITLE
Revisit StructuredDatasetDecoder interface

### PR DESCRIPTION
**Pull request by pingsutw**
Signed-off-by: Kevin Su <pingsutw@apache.org>

# TL;DR
- Add `current_task_metadata` in decoder interface
- Support spark decoder to read subset column
- Remove duplicate test

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2099

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by pingsutw** [comment]
Will fix the tests

File: flytekit/models/literals.py [hunk 1]
```diff
@@ -549,7 +549,7 @@ def from_flyte_idl(cls, pb2_object):
 
 
 class StructuredDatasetMetadata(_common.FlyteIdlEntity):
-    def __init__(self, structured_dataset_type: StructuredDatasetType = None):
```
Comment (wild-endeavor) [inline flytekit/models/literals.py:552]: `Optional[StructuredDatasetType]`?  does the idl translation stuff work okay if it's None?

File: flytekit/types/structured/basic_dfs.py [hunk 2]
```diff
@@ -55,14 +55,13 @@ def decode(
         self,
         ctx: FlyteContext,
         flyte_value: literals.StructuredDataset,
+        current_task_metadata: typing.Optional[StructuredDatasetMetadata] = None,
```
Comment (wild-endeavor) [inline flytekit/types/structured/basic_dfs.py:58]: can we make this not optional?  we can always make one right?  Or am i missing something?

File: flytekit/types/structured/structured_dataset.py [hunk 3]
```diff
@@ -261,14 +261,17 @@ def decode(
...
     ) -> Union[DF, Generator[DF, None, None]]:
         """
         This is code that will be called by the dataset transformer engine to ultimately translate from a Flyte Literal
         value into a Python instance.
 
-        :param ctx:
+        :param ctx: A FlyteContext, useful in accessing the filesystem and other attributes
         :param flyte_value: This will be a Flyte IDL StructuredDataset Literal - do not confuse this with the
           StructuredDataset class defined also in this module.
+        :param current_task_metadata: Metadata contains column name and type, and decoder will use it to
```
Comment (wild-endeavor) [inline flytekit/types/structured/structured_dataset.py:273]: Metadata object containing the type (and columns if any) for the currently executing task. This type may have more or less information than the type information bundled inside the incoming flyte_value.

File: plugins/flytekit-data-fsspec/flytekitplugins/fsspec/pandas.py [hunk 4]
```diff
@@ -58,14 +58,13 @@ def decode(
...
+        current_task_metadata: typing.Optional[StructuredDatasetMetadata] = None,
     ) -> pd.DataFrame:
         uri = flyte_value.uri
         columns = None
         kwargs = get_storage_options(uri)
-        if flyte_value.metadata.structured_dataset_type.columns:
-            columns = []
-            for c in flyte_value.metadata.structured_dataset_type.columns:
-                columns.append(c.name)
+        if current_task_metadata and current_task_metadata.structured_dataset_type.columns:
```
Comment (wild-endeavor) [inline plugins/flytekit-data-fsspec/flytekitplugins/fsspec/pandas.py:66]: If there's no columns here then columns in the read_parquet function will be `None`.  Will that work okay?  It'll pull all columns right?

File: flytekit/models/literals.py (hunk 1 above)
Comment (pingsutw) [reply flytekit/models/literals.py:552]: updated it

File: flytekit/types/structured/basic_dfs.py (hunk 2 above)
Comment (pingsutw) [reply flytekit/types/structured/basic_dfs.py]: you're right, `StructuredDatasetMetadata` will never be None, thanks

File: flytekit/types/structured/structured_dataset.py (hunk 3 above)
Comment (pingsutw) [reply flytekit/types/structured/structured_dataset.py]: tyty

File: plugins/flytekit-data-fsspec/flytekitplugins/fsspec/pandas.py (hunk 4 above)
Comment (pingsutw) [reply plugins/flytekit-data-fsspec/flytekitplugins/fsspec/pandas.py:66]: yeah, will read all the column if `columns` value is None

File: plugins/flytekit-data-fsspec/flytekitplugins/fsspec/arrow.py [hunk 5]
```diff
@@ -47,17 +47,16 @@ def decode(
...
         if not ctx.file_access.is_remote(uri):
             Path(uri).parent.mkdir(parents=True, exist_ok=True)
         _, path = split_protocol(uri)
 
         columns = None
-        if flyte_value.metadata.structured_dataset_type.columns:
-            columns = []
-            for c in flyte_value.metadata.structured_dataset_type.columns:
-                columns.append(c.name)
+        if current_task_metadata and current_task_metadata.structured_dataset_type.columns:
```
Comment (wild-endeavor) [inline plugins/flytekit-data-fsspec/flytekitplugins/fsspec/arrow.py:58]: this is no longer optional right?  Can we remove the first if check?

File: plugins/flytekit-spark/tests/test_wf.py [hunk 6]
```diff
@@ -53,27 +58,6 @@ def my_wf() -> my_schema:
     assert df2 is not None
 
 
-def test_ddwf1_with_spark():
```
Comment (wild-endeavor) [inline plugins/flytekit-spark/tests/test_wf.py:56]: was this test bad for some reason?

File: flytekit/types/structured/basic_dfs.py [hunk 7]
```diff
@@ -55,14 +55,13 @@ def decode(
...
+        current_task_metadata: StructuredDatasetMetadata,
     ) -> pd.DataFrame:
         path = flyte_value.uri
         local_dir = ctx.file_access.get_random_local_directory()
         ctx.file_access.get_data(path, local_dir, is_multipart=True)
-        if flyte_value.metadata.structured_dataset_type.columns:
-            columns = []
-            for c in flyte_value.metadata.structured_dataset_type.columns:
-                columns.append(c.name)
+        if current_task_metadata and current_task_metadata.structured_dataset_type.columns:
```
Comment (wild-endeavor) [inline flytekit/types/structured/basic_dfs.py:63]: same here, do we still need "if current_task_metadata"

**Comment by wild-endeavor** [comment]
just a couple more questions.

File: plugins/flytekit-spark/tests/test_wf.py (hunk 6 above)
Comment (pingsutw) [reply plugins/flytekit-spark/tests/test_wf.py:56]: We have the same test here. https://github.com/flyteorg/flytekit/blob/eb1c75e4fe6875640560a1d6dbdfb377c6c8e494/plugins/flytekit-spark/tests/test_wf.py#L15-L33****

File: plugins/flytekit-data-fsspec/flytekitplugins/fsspec/arrow.py (hunk 5 above)
Comment (pingsutw) [reply plugins/flytekit-data-fsspec/flytekitplugins/fsspec/arrow.py:58]: yeah, updated it to 
```python
if current_task_metadata.structured_dataset_type and current_task_metadata.structured_dataset_type.columns
```

**Review by wild-endeavor — APPROVED** [review]

**Pull request merged**